controllers/storageclassclaimsecret/storageclassclaimsecret_controller.go

```diff
@@ -0,0 +1,691 @@
+			return reconcile.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClassClaim Secret.")
+		return reconcile.Result{}, err
+	}
+
+	r.storageClassClaim = &v1alpha1.StorageClassClaim{}
+	if unmarshalErr := json.Unmarshal(r.storageClassClaimSecret.Data["StorageClassClaim"], r.storageClassClaim); unmarshalErr != nil {
+		r.log.Error(unmarshalErr, "Failed to unmarshal StorageClassClaim Secret.")
+		return reconcile.Result{}, unmarshalErr
+	}
+
+	if r.storageClassClaim.Status.Phase == "" {
+		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimInitializing
+	}
+
+	storageClusterList := &v1.StorageClusterList{}
+	if err := r.list(storageClusterList, client.InNamespace(r.OperatorNamespace)); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch l := len(storageClusterList.Items); {
+	case l == 0:
+		return reconcile.Result{}, fmt.Errorf("no StorageCluster found")
+	case l != 1:
+		return reconcile.Result{}, fmt.Errorf("multiple StorageCluster found")
+	}
+	r.storageCluster = &storageClusterList.Items[0]
+
+	var result reconcile.Result
+	var reconcileError error
+	if storagecluster.IsOCSConsumerMode(r.storageCluster) {
+		return reconcile.Result{}, fmt.Errorf("StorageClassClaim Secret is not supported in OCSConsumer mode")
+	}
+
+	result, reconcileError = r.reconcilePhases()
+
+	if claimBytes, marshalErr := json.Marshal(r.storageClassClaim); marshalErr == nil {
+		r.storageClassClaimSecret.Data["StorageClassClaim"] = claimBytes
+	} else {
+		r.log.Error(marshalErr, "Failed to marshal StorageClassClaim Secret.")
+		return reconcile.Result{}, marshalErr
+	}
+
+	// Apply status changes to the StorageClassClaim
+	statusError := r.update(r.storageClassClaimSecret)
+	if statusError != nil {
+		r.log.Info("Failed to update StorageClassClaim Secret status.")
+	}
+
+	// Reconcile errors have higher priority than status update errors
+	if reconcileError != nil {
+		return result, reconcileError
+	}
+
+	if statusError != nil {
+		return result, statusError
+	}
+
+	return result, nil
+}
+
+func (r *StorageClassClaimSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	enqueueStorageConsumerRequest := handler.EnqueueRequestsFromMapFunc(
+		func(obj client.Object) []reconcile.Request {
+			annotations := obj.GetAnnotations()
+			if annotation, found := annotations[v1alpha1.StorageClassClaimAnnotation]; found {
+				parts := strings.Split(annotation, "/")
+				return []reconcile.Request{{
+					NamespacedName: types.NamespacedName{
+						Namespace: parts[0],
+						Name:      parts[1],
+					},
+				}}
+			}
+			return []reconcile.Request{}
+		})
+
+	// TODO: Annotations or labels...?
+	claimAnnotationPredicate := predicate.NewPredicateFuncs(
+		func(obj client.Object) bool {
+			annotations := obj.GetAnnotations()
+			_, found := annotations[v1alpha1.StorageClassClaimAnnotation]
+			return found
+		},
+	)
+	metaPredicate := predicate.Or(
+		predicate.GenerationChangedPredicate{},
+		util.MetadataChangedPredicate{},
```

Comment:
I believe that metadata changes are captured via generation change. No need to Or against it.

Reply:
Not always! It seems to vary from Kind to Kind (because of course it does!! :sob: ) so I figure it's safe enough to leave it for now even if it's slightly inefficient. Thus far I haven't seen enough log spam to indicate a problem. Might be different when we switch to ConfigMaps, so I might be able to look more deeply into it for that.